src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1636,16 +1671,38 @@ public override BoundNode VisitLocalDeclaration(BoundLocalDeclaration node)
                 }
 
                 type = valueType.ToTypeWithAnnotations();
-                _variableTypes[local] = type;
+                updateLocalSymbols(type);
 
                 if (node.DeclaredTypeOpt != null)
                 {
                     SetAnalyzedNullability(node.DeclaredTypeOpt, new VisitResult(valueType, type), true);
                 }
             }
+            else
+            {
+                updateLocalSymbols();
+            }
 
             TrackNullableStateForAssignment(initializer, type, slot, valueType, MakeSlot(initializer));
             return null;
+
+            void updateLocalSymbols(TypeWithAnnotations? updatedType = null)
+            {
+                var updatedContainingSymbol = GetUpdatedContainingSymbol(node.LocalSymbol);
+                if (Symbol.Equals(node.LocalSymbol.ContainingSymbol, updatedContainingSymbol, TypeCompareKind.ConsiderEverything) &&
+                    (updatedType is null || node.LocalSymbol.TypeWithAnnotations.Equals(updatedType.Value, TypeCompareKind.ConsiderEverything)))
```

Comment:
>node.LocalSymbol.TypeWithAnnotations.Equals(updatedType.Value, TypeCompareKind.ConsiderEverything) [](start = 44, length = 98)

Should we call `SetLocalType()` unconditionally, in case we're updating the type a second time, back to the original value?

Reply:
No longer relevant.

---
In reply to: [327874320](https://github.com/dotnet/roslyn/pull/38769#discussion_r327874320) [](ancestors = 327874320)